### ISSUE TITLE
[BUG] Print head homing after print (with Video)

### ISSUE THREAD

**Issue by derenma** [prusa3d/Prusa-Firmware-Buddy#864]
**Printer type** - MINI

**Printer firmware version**-  4.2.1

**Original or Custom firmware** - Original

**Optional upgrades** - Filament Runout Sensor, Bondtech heatbreak/extruder

**USB drive or USB/Octoprint** - USB

**Describe the bug**
Printer initiates home after print. Edit: This is dissimilar to [#733](https://github.com/prusa3d/Prusa-Firmware-Buddy/issues/733) as that my printer does not initiate a purge sequence.

**How to reproduce**
**_Edit: Found it!! Maybe!_** Attaching video, dump.bin and gcode.

There is a parsing error with the commented `; start_gcode` comment at the bottom of the gcode. Since that is the only other place in the gcode that contains a G28, it must be getting parsed incorrectly by the firmware. This is likely the cause of #733 where only some of the purge sequence was initiated as well.

End gcode comments:
`; start_gcode = ; M301 P13.54 I0.98 D46.58 ; Set new PID values for BondTech heatbreak\nG90 ; use absolute coordinates\nM83 ; extruder relative mode\nM104 S170 ; set extruder temp for bed leveling\nM140 S[first_layer_bed_temperature] ; set bed temp\nM109 R170 ; wait for bed leveling temp\nM190 S[first_layer_bed_temperature] ; wait for bed temp\nG28 ; home all without mesh bed level\nG29 ; mesh bed leveling \nM104 S[first_layer_temperature] ; set extruder temp\nG92 E0.0\nG1 Y-2.0 X179 F2400\nG1 Z3 F720\nM109 S[first_layer_temperature] ; wait for extruder temp\n\nM92 E415; e-step update\n\n; intro line\nG1 X170 F1000\nG1 Z0.2 F720\nG1 X110.0 E8.0 F900\nG1 X40.0 E10.0 F700\nG92 E0.0\n\nM221 S95 ; set flow`

I believe there is a one-off error somewhere in `gcode_file.cpp` when it sees a ton of comments in one line which causes some of the commented gcode to get processed. Take that with a grain of salt as I do not know the code structure that well in this case.

I will test this **_theory_** extensively today.

**Expected behavior**
The printer does not attempt to self-destruct.

**G-code**
Attached.

**Crash dump file**
Attached.

**Video**
 This is a reproduced error that I was able to get on video the second time during the same power cycle: https://youtu.be/E0-OvU_h4do

**Theory/Debugging**
~~I did one manual homing during this power cycle. I will reboot the printer and update this post if there are any changes in the next test print.~~ _**Edit: Testing this theory did nothing.**_

**gcode and dump.bin**
[gcode_and_dump.zip](https://github.com/prusa3d/Prusa-Firmware-Buddy/files/5172747/gcode_and_dump.zip)

**gcode modifiers in slicer**
M92 E415
M301 P13.54 I0.98 D46.58

**current testing iterations w/ reboots between tests**
1. M301 set in memory and in gcode causes failure
2. M301 in memory and **_not_** in gcode causes failure (it was actually in the gcode, just commented out in the slicer)
3. No changes to M301 after reset causes failure (the M301 was still only commented out in the slicer)
4. Re-added (uncommented) M301 manually to the start gcode and removed both `; end_gcode` and `; start_gcode` at the bottom of the file. No failure.

**Proof and replication**

Hmm. I was wrong. The G28 wasn't getting read, but rather the `G1 Y-2.0 X179 F2400`.

The Bad:
`; start_gcode = ; M301 P13.54 I0.98 D46.58 ; Set new PID values for BondTech heatbreak\nG90 ; use absolute coordinates\nM83 ; extruder relative mode\nM104 S170 ; set extruder temp for bed leveling\nM140 S[first_layer_bed_temperature] ; set bed temp\nM109 R170 ; wait for bed leveling temp\nM190 S[first_layer_bed_temperature] ; wait for bed temp\nG28 ; home all without mesh bed level\nG29 ; mesh bed leveling \nM104 S[first_layer_temperature] ; set extruder temp\nG92 E0.0\nG1 Y-2.0 X179 F2400\nG1 Z3 F720\nM109 S[first_layer_temperature] ; wait for extruder temp\n\nM92 E415; e-step update\n\n; intro line\nG1 X170 F1000\nG1 Z0.2 F720\nG1 X110.0 E8.0 F900\nG1 X40.0 E10.0 F700\nG92 E0.0\n\nM221 S95 ; set flow`

The proof [changed `G1 Y-2.0 X179 F2400` to `G1 Y180 X179 F2400`:
`; start_gcode = ; M301 P13.54 I0.98 D46.58 ; Set new PID values for BondTech heatbreak\nG90 ; use absolute coordinates\nM83 ; extruder relative mode\nM104 S170 ; set extruder temp for bed leveling\nM140 S[first_layer_bed_temperature] ; set bed temp\nM109 R170 ; wait for bed leveling temp\nM190 S[first_layer_bed_temperature] ; wait for bed temp\nG28 ; home all without mesh bed level\nG29 ; mesh bed leveling \nM104 S[first_layer_temperature] ; set extruder temp\nG92 E0.0\nG1 Y180 X179 F2400\nG1 Z3 F720\nM109 S[first_layer_temperature] ; wait for extruder temp\n\nM92 E415; e-step update\n\n; intro line\nG1 X170 F1000\nG1 Z0.2 F720\nG1 X110.0 E8.0 F900\nG1 X40.0 E10.0 F700\nG92 E0.0\n\nM221 S95 ; set flow`

The fix. Remove the start semicolon and the comment from the above`  ; M301 P13.54 I0.98 D46.58 ; Set new PID values for BondTech heatbreak\n`:

`; start_gcode = M301 P13.54 I0.98 D46.58\nG90 ; use absolute coordinates\nM83 ; extruder relative mode\nM104 S170 ; set extruder temp for bed leveling\nM140 S[first_layer_bed_temperature] ; set bed temp\nM109 R170 ; wait for bed leveling temp\nM190 S[first_layer_bed_temperature] ; wait for bed temp\nG28 ; home all without mesh bed level\nG29 ; mesh bed leveling \nM104 S[first_layer_temperature] ; set extruder temp\nG92 E0.0\nG1 Y-2.0 X179 F2400\nG1 Z3 F720\nM109 S[first_layer_temperature] ; wait for extruder temp\n\nM92 E415; e-step update\n\n; intro line\nG1 X170 F1000\nG1 Z0.2 F720\nG1 X110.0 E8.0 F900\nG1 X40.0 E10.0 F700\nG92 E0.0\n\nM221 S95 ; set flow`

I have just tested breaking the first line comments, with the M301 line just below, which did not replicate the problem.
`; BREAK ME` and `; BREAK ME ; HERE` as the first lines in the gcode did not replicate the original issue. Weird.

Last Edit:

**Conclusion**
There is a pointer that is getting bumped ahead somehow which is losing track of where '='s and ';'s happen. _I think._ It looks like the firmware is processing the `; start_gcode  = xxx ;` in some way, but that is just weird.

Additional Gcode:
[CaliCube_0.25mm_PETG_MINI_2m_Bad_and_Good.zip](https://github.com/prusa3d/Prusa-Firmware-Buddy/files/5201814/CaliCube_0.25mm_PETG_MINI_2m_Bad_and_Good.zip)

**Comment by neophyl** [comment]
#733 ?

**Comment by derenma** [comment]
@neophyl Possibly. I am not getting the purge sequence, just the home.

Doing a few more tests today to see if I can isolate the issue. This seems like it would be a pain for Prusa to replicate without me sending my whole machine back to them. ;)

I am really thinking it has to do with the M301 that I just added since that is the only notable change that I have done in the last day. There are a few scenarios I can still test and will reconnect my OctoPrint pi to see if the console will tell me anything.

**Comment by derenma** [comment]
@JohnnyDeer This is a difficult one to replicate, so godspeed.  If you want the full "good" and "bad" gcode, let me know. (I unfortunately had Prusament Orange loaded printing 20x test cubes, so that sucked. ;) )

**Comment by JohnnyDeer** [comment]
Hi @derenma yes please, it would be helpful. We think this issue may have same reason as #733.

**Comment by derenma** [comment]
@JohnnyDeer Attached the two gcode files at the end of the problem description. Once I was confident that I found the problem, I changed the test cube to around 5mm^3 and it had no effect on the underlying problem. These super-quick prints were printed in Prusament PETG.

**Comment by derenma** [comment]
The error has to be in here somewhere: https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/e2445b148f679a893ddc33f1d6f773d1e07c115f/src/common/gcode_file.cpp#L94-L124

'=' is still getting processed after the comments. Totally cool, as I have no idea what ya'll are doing with the comments. I am thinking there is an exception issue, which is total speculation.

**Comment by derenma** [comment]
@JohnnyDeer I am curious if you are able to replicate. Please, let me know.

**Comment by derenma** [comment]
If not, it points to a hardware fault. It you can do it, it's firmware.

**Comment by danielo515** [comment]
I am also having a similar behavior

**Comment by zoltan-l** [comment]
derenma, did you try to remove the M92 E415 G-code from your .gcode file?

**Comment by derenma** [comment]
@zoltan-l I did not. I have to admit that is a good troubleshooting step that I overlooked, but setting esteps for the Bondtech extruder is super critical, IMHO. I'll try this tomorrow in the morning.

**Comment by zoltan-l** [comment]
@derenma 
thanks, looking forward to your response

**Comment by derenma** [comment]
@zoltan-l It's taking me a bit to revert back to a broken state. :) Should be there soon.

**Comment by zoltan-l** [comment]
I miss any video proving all have the same issue. I could reproduce the issue using (almost ) identical G-code with and without M92

**Comment by derenma** [comment]
@zoltan-l Tried the following:

1. Commenting out the M92 at the top of the code - Printer failed.
2. Deleting the M92 completely from the top of the code - Printer failed.
3. Attempted a NOP slide into `; start_gcode = AAAAAA...` - Print worked as expected. (I have no idea what the stack looks like for this control board, or if I could even attempt such a hack.  That was more just for lulz.)

**Comment by derenma** [comment]
Since deleting two lines of gcode above `; start_gcode =` had no effect so the problem is isolated, I believe, to how the comments are getting parsed at the end of the gcode file on the printer itself.

I have posted and deleted a few wild theories in this thread about this, but alas, they were not constructive in hindsight.

**Comment by derenma** [comment]
@zoltan-l Oh S***. I went to start a print just now and the NOP slide actually worked somehow! I thought the last print test ended in a "complete" state, but the printer actually paused after the print. It didn't actually complete!

This opens some doors for exploration, that's for sure.

**Comment by derenma** [comment]
[ deleted crazy engineer ranting here ]

**Comment by zoltan-l** [comment]
The cause is the same like https://github.com/prusa3d/Prusa-Firmware-Buddy/issues/733
workaround same - remove from Gcode the long ( longer than 65 characters) comment line(s) which is (are) misinterpreted by parser.

**Comment by derenma** [comment]
@zoltan-l Wheeeeeee! I hope I was able to provide enough breadcrumbs for you to work with. Cheers!

**Comment by zoltan-l** [comment]
@derenma 
Matt thanks, you were great !

**Comment by derenma** [comment]
Woohoo!  Just saw the recent commit to fix this issue and it looks solid, FWIW.

**Comment by derenma** [comment]
LGTM. Ship it.

**Comment by derenma** [comment]
(not a fan of statically defining issues, but that is filtering 101.)

**Comment by JohnnyDeer** [comment]
New FW version with fix of this bug is out. @derenma please try it out ad let know if this issue can be closed :)
https://github.com/prusa3d/Prusa-Firmware-Buddy/releases/tag/v4.3.2-RC1

**Comment by derenma** [comment]
@JohnnyDeer @zoltan-l BAM! Put a bow on it and ship it. The "bad" g-codeI posted a few months ago prints just fine under the exact same conditions.

For lulz, I just used a garbage USB drive and a USB extension that is possibly going bad. I had suspected there were additional USB issues that _might_ have been part of this problem, but wasn't sure if there was an association or not. I think I deleted some speculative rants about that in the past.....
